testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketGatheringWriteTest.java

```diff
@@ -175,49 +166,84 @@ private void testGatheringWrite0(
+        private final Promise<Void> doneReadingPromise;
+        final ByteBuf received = Unpooled.buffer();
+
+        TestServerHandler(boolean autoRead, Promise<Void> doneReadingPromise, int expectedBytes) {
+            super(autoRead);
+            this.doneReadingPromise = doneReadingPromise;
+            this.expectedBytes = expectedBytes;
+        }
+
+        @Override
+        public void channelRead0(ChannelHandlerContext ctx, ByteBuf in) throws Exception {
+            received.writeBytes(in);
+            if (received.readableBytes() >= expectedBytes) {
+                doneReadingPromise.setSuccess(null);
+            }
+        }
+
+        @Override
+        void handleException(ChannelHandlerContext ctx, Throwable cause) {
+            doneReadingPromise.tryFailure(cause);
+            super.handleException(ctx, cause);
+        }
+
+        @Override
+        public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+            doneReadingPromise.tryFailure(new IllegalStateException("server closed!"));
+            super.channelInactive(ctx);
+        }
     }
 
     private static class TestHandler extends SimpleChannelInboundHandler<ByteBuf> {
         private final boolean autoRead;
         volatile Channel channel;
         final AtomicReference<Throwable> exception = new AtomicReference<Throwable>();
-        volatile int counter;
-        final ByteBuf received = Unpooled.buffer();
 
         TestHandler(boolean autoRead) {
             this.autoRead = autoRead;
         }
 
         @Override
-        public void channelActive(ChannelHandlerContext ctx)
-                throws Exception {
+        public final void channelActive(ChannelHandlerContext ctx) throws Exception {
             channel = ctx.channel();
             if (!autoRead) {
                 ctx.read();
             }
+            super.channelInactive(ctx);
         }
 
         @Override
         public void channelRead0(ChannelHandlerContext ctx, ByteBuf in) throws Exception {
-            counter += in.readableBytes();
-            received.writeBytes(in);
         }
 
         @Override
-        public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+        public final void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
             if (!autoRead) {
                 ctx.read();
             }
+            super.channelReadComplete(ctx);
         }
 
         @Override
-        public void exceptionCaught(ChannelHandlerContext ctx,
-                Throwable cause) throws Exception {
+        public final void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
             if (exception.compareAndSet(null, cause)) {
-                ctx.close();
+                handleException(ctx, cause);
             }
+            super.exceptionCaught(ctx, cause);
+        }
+
+        void handleException(ChannelHandlerContext ctx, Throwable cause) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "cause". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)